### DISCUSSION TITLE
Execution sequence correct?

### DISCUSSION THREAD

**Discussion by mizamae** [django/channels#1728] [Q&A]
I am trying to execute a group_send sequentially within a Django view. The synchronous function from the model that is being called from the view is:

    def findNewDevice():
        from MQTT.constants import MQTT_TOPIC_AUTODISCOVER
        cache.set(key='scanning',value=True,timeout=10)
        
        from channels.layers import get_channel_layer
        channel_layer = get_channel_layer()
        from asgiref.sync import async_to_sync
        from MainAPP.consumers import SystemConsumer
        async_to_sync(channel_layer.group_send)("MainAPPSystem", {"type": "publish.ScanningStatus"})
        
        from time import sleep
        DEVICE.newDevice_FIFO=queue.Queue(maxsize=1)
        i=0
        while DEVICE.newDevice_FIFO.empty() and i<=5:
            SignalPublishMQTTTopic.send(sender='DEVICE',TOPIC=MQTT_TOPIC_AUTODISCOVER,VALUE=0,QOS=2)
            i=i+1
            async_to_sync(channel_layer.group_send)("MainAPPSystem", {"type": "publish.ScanningStatus"})
            sleep(1)

When I enter into this findNewDevice function, it reaches the line with:

    async_to_sync(channel_layer.group_send)("MainAPPSystem", {"type": "publish.ScanningStatus"})

and it continues the execution without entering into the publish_ScanningStatus of the consumer.

    class SystemConsumer(JsonWebsocketConsumer):
        groups = ["Broadcast","MainAPPSystem"]
    
        def connect(self):
            self.accept()
        
        def publish_ScanningStatus(self,event):
            from django.core.cache import cache
            scanning=cache.get(key='scanning',default=False)
            self.send_json({"action":"scanning_status",
                          "value":scanning,
                          })
            print("Published scanning status: " +scanning)

after it finishes (5 seconds), it then enters the publish_ScanningStatus of the consumer only once. I thought that by including async_to_sync into a function, it converted it into sequential code, so it should execute the function instantaneously; obviously I understood wrongly so is there anything I could do to get a sequential execution of the code??

What I would need to have is send the message to the group (to the client) so that I can indicate it that the system is scanning for new devices...

Thanks!!

**Comment by yedpodtrzitko** [comment]
Looks like there's nothing asynchronous in the method `publish_ScanningStatus`, so either you dont need `async_to_sync` or you forgot to add `await` there (presumably to `self.send_json()`)?
